cpp/src/arrow/compute/exec.h

```diff
@@ -131,26 +131,60 @@ class ARROW_EXPORT ExecContext {
 /// implementations. This is especially relevant for aggregations but also
 /// applies to scalar operations.
 ///
-/// We are not yet using this so this is mostly a placeholder for now.
-///
 /// [1]: http://cidrdb.org/cidr2005/papers/P19.pdf
 class ARROW_EXPORT SelectionVector {
  public:
   explicit SelectionVector(std::shared_ptr<ArrayData> data);
 
   explicit SelectionVector(const Array& arr);
 
-  /// \brief Create SelectionVector from boolean mask
-  static Result<std::shared_ptr<SelectionVector>> FromMask(const BooleanArray& arr);
-
+  std::shared_ptr<ArrayData> data() const { return data_; }
   const int32_t* indices() const { return indices_; }
-  int32_t length() const;
+  int64_t length() const;
+
+  Status Validate(int64_t values_length = -1) const;
 
  private:
   std::shared_ptr<ArrayData> data_;
   const int32_t* indices_;
 };
 
+/// \brief A span of a SelectionVector's indices. Can represent a slice of the
+/// underlying indices.
+///
+/// Note that as an indirection of indices to the data in an ExecBatch, when sliced
+/// along with the batch, the indices themselves need to be back-shifted to be relative to
+/// the batch slice (ExecSpan). For example, consider an ExecBatch of 10 rows with a
+/// SelectionVector [0, 1, 9] is to be executed per-8-rows. The first slice of the batch
+/// will have row 0 to row 7 of the original batch with selection slice [0, 1]. The second
+/// slice of the batch will have row 8 and row 9 of the original batch however they are
+/// referred to as row 0 and row 1 by the kernel. Therefore the second selection slice
+/// should be [9 - 8] = [1]. This is done by setting index_back_shift to 8 for the second
+/// selection slice.
+class ARROW_EXPORT SelectionVectorSpan {
+ public:
+  explicit SelectionVectorSpan(const int32_t* indices = NULLPTR, int64_t length = 0,
+                               int64_t offset = 0, int32_t index_back_shift = 0)
+      : indices_(indices),
+        length_(length),
+        offset_(offset),
+        index_back_shift_(index_back_shift) {}
+
+  void SetSlice(int64_t offset, int64_t length, int32_t index_back_shift = 0);
+
+  int32_t operator[](int64_t i) const {
+    return indices_[i + offset_] - index_back_shift_;
```

Comment:
When you use this class in loops, you will probably get better assembly if it's copied into a local variable (in the "stack") before the loop to get SROA [1] to kick in and then you can keep all these members in registers.

[1] https://blog.regehr.org/archives/1603